src/PowerShell/Help/Microsoft.WinGet.Client/Assert-WinGetPackageManager.md

```diff
@@ -0,0 +1,125 @@
+online version:
+schema: 2.0.0
+---
+
+# Assert-WinGetPackageManager
+
+## SYNOPSIS
+Verifies WinGet is installed properly.
+
+## SYNTAX
+
+### IntegrityVersionSet (Default)
+```
+Assert-WinGetPackageManager [-Version <String>] [<CommonParameters>]
+```
+
+### IntegrityLatestSet
+```
+Assert-WinGetPackageManager [-Latest] [-IncludePreRelease] [<CommonParameters>]
+```
+
+## DESCRIPTION
+Verifies WinGet is installed properly.
+
+>**This does not ensure the latest version of WinGet is installed. This just verifies the current installed version correlated with the Microsoft.WinGet.Client module is functional.**
+
+## EXAMPLES
+
+### Example 1: Default usage
+
+```powershell
+PS C:\> Assert-WinGetPackageManager
+```
+
+If the current version of WinGet is installed correctly this will return without error.
+
+### Example 2: Check if latest stable version is installed
+
+```powershell
+PS C:\> Assert-WinGetPackageManager -Latest
+```
+
+If the latest version of WinGet correlated with the Microsoft.WinGet.Client module is installed correctly this will return without error.
+
+### Example 3: Check if latest preview version is installed
+
+```powershell
+PS C:\> Assert-WinGetPackageManager -IncludePreRelease
+```
+
+If a preview version of WinGet correlated with the Microsoft.WinGet.Client module is installed correctly this will return without error.
+
+### Example 4: Check if specific version is installed
+
+```powershell
+PS C:\> Assert-WinGetPackageManager -Version v1.8.1911
+```
+
+If the specified version of WinGet correlated with the Microsoft.WinGet.Client module is installed correctly this will return without error.
+
+## PARAMETERS
+
+### -IncludePreRelease
+Includes preview versions of WinGet
+
+```yaml
+Type: System.Management.Automation.SwitchParameter
+Parameter Sets: IntegrityLatestSet
+Aliases:
+
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
+
+### -Latest
+Verifies the latest version of WinGet correlated with the Microsoft.WinGet.Client module is installed
+
+```yaml
+Type: System.Management.Automation.SwitchParameter
+Parameter Sets: IntegrityLatestSet
+Aliases:
+
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
+
+### -Version
+Verifies a specific version of WinGet is installed.
+
+```yaml
+Type: System.String
+Parameter Sets: IntegrityVersionSet
+Aliases:
+
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: True (ByPropertyName)
+Accept wildcard characters: False
+```
+
+### CommonParameters
+This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable, -InformationAction, -InformationVariable, -OutVariable, -OutBuffer, -PipelineVariable, -ProgressAction, -Verbose, -WarningAction, and -WarningVariable. For more information, see [about_CommonParameters](http://go.microsoft.com/fwlink/?LinkID=113216).
```

Comment:
The current version of `Assert-WinGetPackageManager` (1.8.1911) actually throws when I try to use the `-Debug` parameter:

```
PS C:\Windows\system32> Assert-WinGetPackageManager -Debug
Assert-WinGetPackageManager : Debug parameter not supported
At line:1 char:1
+ Assert-WinGetPackageManager -Debug
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Assert-WinGetPackageManager], NotSupportedException
    + FullyQualifiedErrorId : System.NotSupportedException,Microsoft.WinGet.Client.Commands.AssertWinGetPackageManagerCmdlet
```

That's unfortunate cmdlet design because it's unexpected / unprecedented in PowerShell. Even PlatyPS as we can see just assumes the common parameters are either implemented or at least available and just not utilized, but a cmdlet really shouldn't throw an exception when attempting to use them. The way it is now, `-Debug` would have to be removed from this list of supported parameters because it's clearly not supported, but really the cmdlet should be updated to just not throw that exception.

Reply:
https://github.com/microsoft/winget-cli/issues/4697

Reply:
Thanks for the issue. In case these docs release before that's addressed:

```suggestion
### CommonParameters
This cmdlet supports the common parameters: -ErrorAction, -ErrorVariable, -InformationAction, -InformationVariable, -OutVariable, -OutBuffer, -PipelineVariable, -ProgressAction, -Verbose, -WarningAction, and -WarningVariable. For more information, see [about_CommonParameters](http://go.microsoft.com/fwlink/?LinkID=113216).
```

Reply:
Do you think I should remove "-Debug" from all the cmdlets or is something like the output from:
`Add-WinGetSource -Name mysource -Argument https://contoso.com/ -Type Microsoft.Rest -Debug` reasonable?

```PowerShell
PS C:\Users\denelon> Add-WinGetSource -Name mysource -Argument https://contoso.com/ -Type Microsoft.Rest -Debug
Add-WinGetSource: Debug parameter not supported
```

Reply:
Most importantly that output shouldn't go to the error stream like it does now. Typically, a function / cmdlet implemented in the PowerShell language itself would just accept the `-Debug` parameter and then maybe not do anything with it. E.g.:

```powershell
function Test-DebugParameter {
    [CmdletBinding()]
    Param ()

    # Write-Debug "something"
}
```

if you paste this into your terminal to define the function, and then run either `Test-DebugParameter` or `Test-DebugParameter -Debug` you don't get an error, you just get nothing (unless you uncomment the `Write-Debug` line). Because most users of PowerShell also write their functions, modules and scripts in the PowerShell language, this is the expected behavior. You could of course go out of your way to implement an exception throw when the user uses `-Debug` but for what? The only thing that would cause is scripts that have `$ErrorActionPreference` set to `"Stop"` (basically fail-early-mode) to quit when they encounter this exception/error.

The `-Debug` parameter not being implemented yet is not an error, it is actually kind of common in PowerShell functions. Many functions people write only utilize `Verbose` output for example, not the additional detail level of `Debug`.

If you feel like you must expressly inform the user that there's currently no Debug output to be expected despite the parameter being available then that disclaimer, albeit unusual, would belong on the the `Information` stream (imo). Here's some official guidance on how to utilize PowerShells' unusual amount of output streams: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_output_streams?view=powershell-7.4 (most shells of course only have STDOUT (success) and STDERR (for errors, often misused for verbose information as well). Perhaps you put this information in the error stream because of this (coming from C++), but in PowerShell it's really neat that we can more precisely enable/disable or redirect this total of **6** different output streams and don't have to pollute the error stream with non-errors.